pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -18,23 +18,58 @@ package credentials
 
 import (
 	"encoding/base64"
+	"errors"
 	"fmt"
+	"net/http"
 	"strings"
 	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
 	"github.com/aws/aws-sdk-go/aws/request"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/ecr"
 	"k8s.io/klog"
 
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/kubernetes/pkg/credentialprovider"
 )
 
-const awsChinaRegionPrefix = "cn-"
-const awsStandardDNSSuffix = "amazonaws.com"
-const awsChinaDNSSuffix = "amazonaws.com.cn"
-const registryURLTemplate = "*.dkr.ecr.%s.%s"
+const (
+	awsChinaRegionPrefix        = "cn-"
+	awsStandardDNSSuffix        = "amazonaws.com"
+	awsChinaDNSSuffix           = "amazonaws.com.cn"
+	registryURLTemplate         = "*.dkr.ecr.%s.%s"
+	awsOsakaLocalRegion         = "ap-northeast-3"
+	awsAvailabilityZoneEndpoint = "placement/availability-zone"
+)
+
+func init() {
+	seen := sets.NewString()
+
+	for _, p := range endpoints.DefaultPartitions() {
```

Comment:
this results in registering aws credential providers in non-aws and in non-cloud invocations. don't do this directly in an init block

Reply:
`registerCredentialProvider` explicitly says "This should be called only if using the AWS cloud provider."

Reply:
I think the reason for that is because the credential provider doesn't correctly implement the [Enabled](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/aws/aws_credentials.go#L117-L119) method so the provider is always active if registered. I propose instead of doing the check here to fix `Enabled` and leave this alone. WDYT?

Reply:
how many providers would get registered? the other providers register 1-3

when are you proposing to do the check for aws env, in init() or in Enabled()? Enabled() is documented to be potentially blocking, so it seems better to do the check there. If all we are checking is whether we are in an aws env, can we do the check once in a do.Once() call and share the results among all the registered aws providers?

Reply:
One credential provider for region would be registered, which is currently 18. The credential providers are lazy and won't query metadata unless it is necessary so their very existence should not be problematic.

I propose (and have updated the CR to this end) to do the checking in `Enabled` which seems to be the purpose of that API. The check itself is very lightweight, in the worst case reading up to 3 files in `/sys` and only one on modern instances. These sysfs reads should never block. Given that I don't see any point for the added complexity of synchronization and caching of this information.

I could see the case for this logic to be moved to `init` but if we go that route I think all providers should be refactored to do that as well.

Reply:
@liggitt @mcrute any updates on this?

Reply:
I'm just waiting to hear back from @liggitt to resolve this particular thread then overall approval from @justinsb that I've addressed his concerns, then we should be good to go.

Reply:
I'm not as familiar with the credential provider call patterns. Are all registered providers checked with .Enabled() on every image pull, or only once or twice at kubelet startup? If only at kubelet startup, then putting uncached/unshared logic in Enabled() is fine. If every provider is consulted on every image pull, then sharing and caching whether we are in an aws env seems better.

Reply:
I'm not super familiar with credential provider either, but from my understanding of the lazy provider, it will run `Enabled()` if a registry URL matches a pattern set by the provider. So for the AWS case, we register 18 credential providers (1 per region) but we would only check `Enabled()` if an ECR image on that region was being pulled for the first time. In most environments, I imagine this will be only 1 - 3 times every 12 hours (when ECR certs expire). 

Please correct me if I'm wrong @mcrute

Reply:
`Enabled()` gets checked once when the keyring gets created. In the case of the `dockershim` this happens each time a sandbox is created in kubelet. In the case of the generic `kuberuntime_manager` this happens once when the runtime manager is first created. However, the `onAWSInstance` check is very lightweight, only reading, at worst, 3 virtual files from `/sys` so this should have almost no impact at runtime.

I updated the current region check to only happen on AWS instances to avoid the metadata call. This is really just to handle the edge case of a starting the credential provider in a region that's not yet supported in the SDK and should be a rare case.

There's a potential refactor here to allow only registering one credential provider for AWS instead of 18 but it will be a pretty substantial refactor of the way the credential provider works and I'd prefer to ship this and do that in a different review.

Does the revised approach solve the outstanding issues?

Reply:
I'm all in favor of snipping dependencies, but I don't think putting active aws-specific file-reading code in an init() block like this is a good tradeoff.

Reply:
If dockershim is inadequate relative to kuberuntime_manager, why not fix dockershim so that you can put this where it belongs in `Enabled`?

Reply:
> I'm all in favor of snipping dependencies, but I don't think putting active aws-specific file-reading code in an init() block like this is a good tradeoff.

This is fair. Seems like a larger refactor of credential provider is going to be needed for this. Thanks for the input @liggitt! 

> If dockershim is inadequate relative to kuberuntime_manager, why not fix dockershim so that you can put this where it belongs in `Enabled`?

We'll look into this as well. Thanks @smarterclayton!